### ISSUE TITLE
Remove `LOAD_CLOSURE`

### ISSUE THREAD

**Issue by markshannon** [python/cpython#105775]
`LOAD_CLOSURE` is identical to `LOAD_FAST_CHECK` in every way except its name and number.

The justification for its existence is that "We keep LOAD_CLOSURE so that the bytecode stays more readable.".
Which is insufficient justification to keep it given that it:
* Uses an instruction, a limited resource which adds bulk to the interpreter
* Prevents superinstruction formation.
* Prevents removal of checks for uninitialized variables.

### Linked PRs
* gh-106059

**Comment by heysujal** [comment]
Can I work on it ?

**Comment by heysujal** [comment]
I'd like to work on this issue

**Comment by markshannon** [comment]
Are you sure? This is more complicated than it looks as it involves some awkward compiler internals.

**Comment by heysujal** [comment]
Oh, I thought it would be just deleting some statements

**Comment by markshannon** [comment]
It should be. But the compiler handles local variables in a somewhat convoluted way

**Comment by heysujal** [comment]
Ok, then someone else can pick the issue up.

**Comment by markshannon** [comment]
Rather than add special cases  to `_PyCfgInstruction` and `_PyCompile_Instruction` would it make more sense use the names of the variables as operands, rather than indices into two different lists?
The numbering can then be done in the assembler.

There is no reason to treat `LOAD_CLOSURE` differently to `LOAD_FAST` in the optimizer.

**Comment by carljm** [comment]
If you don't want to tackle a full refactoring of how local variable and cell offsets are handled in this PR, a simpler incremental approach would be to make `LOAD_CLOSURE` a pseudo-instruction (similar to e.g. `STORE_FAST_MAYBE_NULL`) which is emitted in codegen (and thus can still be recognized in `fix_cell_offsets`) but is converted to `LOAD_FAST` in `_PyCfg_ConvertPseudoOps`. This would eliminate `LOAD_CLOSURE` as a real instruction that has to be handled at runtime.

The refactor of cell index handling could still be done as a follow-up PR, allowing removal of the `LOAD_CLOSURE` pseudo-op.

**Comment by gvanrossum** [comment]
(From your post to core-mentorship:)

> When I first set up my cpython development environment, I followed the guidance on the developer’s guide and have been studying the compiler design page closely. From my reading of it, when I remove an opcode, I need to
> 
> - Remove references to the opcode from the code and update any logic impacted
> - Remove the opcode from Lib/opcode.py and the dis module documentation
> - Run make regen-opcode regen-opcode-targets to regenerate Include/opcode.h and Python/opcode_targets.h
> - Update the magic number in Lib/importlib/_bootstrap_external.py and PC/launcher.c
> - "Add the new bytecode target" 
> - Run make regen-importlib
> 
> The first step is probably most of the work, but I'm a little bit unclear about the second to last step.
> 
> The compiler design page in the developer guide has written
> 
> > Changes to Lib/importlib/_bootstrap_external.py will take effect only after running make regen-importlib. Running this command before adding the new bytecode target to Python/ceval.c will result in an error. You should only run make regen-importlib after the new bytecode target has been added.
> 
> It makes sense to me that if we’re introducing a new opcode, we should ideally have logic to handle the new instructions in `ceval.c`, but it’s not clear to me what error would be triggered if we’re not referencing the new instruction, so I feel I’m not understanding something. I also don't see most opcodes referenced in ceval.c -- for instance, no LOAD_FAST, STORE_FAST, or JUMP_* instructions, at least not explicitly.

The instructions may well be out of date. (A PR to update them would be welcome of course!)

As of 3.12, the cases in the switch are generated by tooling in Tool/cases_generator (check out the README.md there). This reads Python/bytecodes.c and writes Python/generated_cases.c.h, which in `#include`d by ceval.c in the switch body. **IMPORTANT:** Whenever you edit bytecodes.c, run `make regen-cases` to regenerate the output files.
likely that Carl's suggestion is right: make LOAD_CLOSURE a pseudo op (search for "pseudo" in bytecodes.c) so the compiler can distinguish between it and LOAD_FAST if it needs to, make the assembler translate it to LOAD_FAST. I think that it's totally fine if a sequence of e.g. LOAD_FAST, LOAD_CLOSURE ends up being replaced by a single LOAD_FAST_LOAD_FAST instruction (I think that's what Mark meant).

If you are still stuck, can you push your work to a branch in your GitHub fork of the cpython repo and link it here, and post the errors you are getting (and from which command)?

**Comment by gvanrossum** [comment]
Usually when messing around with the interpreter, once you get it to _build_ it's pretty solid, and if you get it to pass the test suite, it's golden. It seems you've already reached that stage. :-) Platinum would be to pass the buildbots, esp. the refleak subset. IIUC only triagers and core devs can trigger a buildbot run (by setting a magic label on the PR), and will do so as part of the PR review. I can do that for you.

Whether a test for `_PyCfg_ConvertPseudoOps` is needed, maybe @iritkatriel has an opinion.

**Comment by iritkatriel** [comment]
``_PyCfg_ConvertPseudoOps`` can be tested as part of the assembler. See Lib/test/test_compiler_assemble.py, which calls ``_PyCompile_Assemble``. This is pretty new, so there aren't many tests there yet.

**Comment by iritkatriel** [comment]
I wouldn't do that. I'd just add a test to Lib/test/test_compiler_assemble.py that creates a code object from an instruction sequence that contains the pseudo ops, and then checks that the code works as expected.

If the pseudo ops are not replaced the code object can't possibly work.

**Comment by carljm** [comment]
> In terms of possible approaches, @carljm https://github.com/python/cpython/pull/106059#discussion_r1242397712 that we could either refactor the cell/local offsets logic so that we can remove LOAD_CLOSURE altogether, or consider adding explicit LOAD_CLOSURE support to some optimizations

I would say that the former is the ideal forward-looking approach, but may be high difficulty/complexity; it should be possible, but I haven't looked at it in detail. The latter is more of a short-term approach if for performance reasons we urgently want `LOAD_CLOSURE` to participate better in some specific bytecode optimizations, but we haven't been able to achieve the former yet.

**Comment by markshannon** [comment]
I don't think this is urgent.
If it is too awkward to implement now, it can wait.

We plan (vaguely) to change the compiler to use names, rather than indices for load instructions.
Once that change has been made, removing `LOAD_CLOSURE` should be almost trivial.
